File: Chapter 03/Python/section_3_3_4_isp/after/midfielder.py
from .interfaces import FieldPlayerTraining

"""
THE CLEAN IMPLEMENTATION.
* ARCHITECTURE NOTE: The Midfielder now only inherits from FieldPlayerTraining. 
It is no longer "haunted" by goalie methods it can't perform. There is 
no longer a need for NotImplementedErrors, making the class "honest" 
and predictable for any consumer.
"""
class Midfielder(FieldPlayerTraining):
    def practice_shooting(self):
        print("  [Midfielder] Practicing shooting drills.")

    def practice_tackling(self):
        print("  [Midfielder] Practicing slide tackles.")

File: Chapter 03/JS/section_3_3_4_isp/after/midfielder.js
/**
 * THE CLEAN IMPLEMENTATION.
 * * ARCHITECTURE NOTE: Look at how clean this class is now! There are no 
 * Errors being thrown for "diving saves" or "hand distribution." By segregating 
 * the conceptual training roles, the Midfielder only implements what it 
 * actually does. The class is now highly cohesive and "honest."
 */
class Midfielder {
    /**
     * A valid domain action for a field player.
     */
    practiceShooting() {
        console.log("  [Midfielder] Practicing shooting drills.");
    }

    /**
     * A valid domain action for a field player.
     */
    practiceTackling() {
        console.log("  [Midfielder] Practicing slide tackles.");
    }
}

module.exports = Midfielder;